### ISSUE TITLE
Go into insert mode always switches to US keyboard

### ISSUE THREAD

**Issue by erikbra** [aioutecism/amVim-for-VSCode#23]
I am using amVim with Norwegian keyboard layout. Whenever I press "i", i get US keyboard mapping. But, after moving around, the Norwegian keyboard layout is restored. 

How can I help debugging this error?

**Comment by lf94** [comment]
That's really interesting, because I use Dvorak and that never happens to me.

**Comment by aioutecism** [comment]
@erikbra 
I can't reproduce this bug on OSX.
What OS do you use?

**Comment by aioutecism** [comment]
It would be helpful if you could upload a screen capture.

Here is a soft to do this: http://www.cockos.com/licecap/

**Comment by erikbra** [comment]
I'm on Windows. I think I found the problem more accurately. 

In navigation mode, there is always English keyboard layout, so no navigation using non-alphanumeric characters work as expected in my keyboard layout (^, $, etc).

Whenever I enter edit mode, every character I type is according to English keyboard layout, *until* I press an alphanumeric character (in English keyboard layout). Then it switches to the correct keyboard layout. I have attached a screen capture and a picture of my keyboard. 

I type *i* to enter edit mode, then I press the following keys on my keyboard:

øæå'¨\aøæå'¨\ (number 1-6 on attached image, then the *a* key, then number 1-6 on attached again)

![dsc_0004](https://cloud.githubusercontent.com/assets/1628994/12288066/4ffdbc9c-b9d1-11e5-940c-0da833bf79b8.JPG)

That is, I press some keys that are all non-alphanumeric in English keyboard layout, then I press the *a* key, and then press the exact same keys again, now the correct characters are entered.

![amvim-vscode-problem](https://cloud.githubusercontent.com/assets/1628994/12288069/58e1f922-b9d1-11e5-9812-45713dfb0c0a.gif)

So, to sum up, in navigation mode, and in edit mode before pressing any alphanumeric characters, English keyboard layout is always used. That is the bug.

**Comment by aioutecism** [comment]
Thanks for the description, very helpful.
The first `']/;[ is not English layout either.

Do you get the same bug without this extension installed?
I'm thinking if this is a bug of VSCode.

**Comment by erikbra** [comment]
The first letters are English layout, yes. Compare with, e.g. [Google Image of English keyboard layout] (  https://www.google.no/search?q=en+us+keyboard+layout&espv=2&biw=1804&bih=1142&tbm=isch&imgil=UJAlroF2bYWYzM%253A%253BFjoXPueeZuaBfM%253Bhttps%25253A%25252F%25252Fen.wikipedia.org%25252Fwiki%25252FKeyboard_layout&source=iu&pf=m&fir=UJAlroF2bYWYzM%253A%252CFjoXPueeZuaBfM%252C_&usg=__DER-COtq8b9p0U3ZsX_dcnvWM-0%3D&ved=0ahUKEwjW64_97KbKAhUK2xoKHcpQCBIQyjcIJA&ei=YEuWVta7CYq2a8qhoZAB#imgrc=UJAlroF2bYWYzM%3A&usg=__DER-COtq8b9p0U3ZsX_dcnvWM-0%3D)

It is not a VSCode bug. Norwegian layout works as expected when I uninstall the extension.
There is no "normal/navigation mode" in vscode without amVim either, so it's really not possible ;)

**Comment by aioutecism** [comment]
Can you try the same thing using https://github.com/VSCodeVim/Vim ?
I want to confirm if it is the problem with VSCode's package.json.

**Comment by erikbra** [comment]
VsCodeVim obviously has a concious choice about this, because it lets me configure (some) keyboard layouts:

*Configure

Adjust configurations through user settings (File -> Preferences -> User Settings).

vim.keyboardLayout:
Supported Values: en-US (QWERTY) (default), es-ES (QWERTY), de-DE (QWERTZ), da-DK (QWERTY)*

I seriously don't understand why a vi input mode plugin should have its own keyboard layout settings, and not use the default in Windows.

VSCodeVim uses English keyboard layout in "normal/navigation" mode (which is annoying enough). However, VSCodeVim switches to my Norwegian keyboard layout at once if I enter edit mode, which is better.

Why does Code in a way or another have its own keyboard layout? is this a bug or a feature? Or have you conciously chosen another keyboard layout in your plugin?

**Comment by aioutecism** [comment]
Extensions can't get layout informations from VSCode. Infact, VSCode should handle these for the extension, but it doesn't for now. 
Please give me some time to solve this problem.

**Comment by erikbra** [comment]
Thanks. Just tell me if you need any more help troubleshooting or testing the issue.

**Comment by erikbra** [comment]
Any progress on this, @aioutecism? Anything I can do to contribute?

**Comment by aioutecism** [comment]
@erikbra 
I'm trying to add a layer to mapping system to deal with all kinds of keyboard layout.
I would be great if you can help test when I finish it.
Please give me some more time.

**Comment by latesr** [comment]
I'd be keen to try the proposed fix as well.  I'm on OSX and appear to have a related issue.  I use a British keyboard mapping where Shift-2 is the quote character: " However, with the amVim plugin installed, when in edit mode, pressing Shift-2 gives the @ symbol.  Further, the @ key gives the ".  This is consistent with an American keyboard, but inconsistent with the British one.

**Comment by aioutecism** [comment]
@erikbra @latesr 

I added keyboard layout support here: #45.

It has only the `US QWERTY` layout now. Could you help creating layout you need?
All it needs is a file like this:
https://github.com/aioutecism/amVim-for-VSCode/blob/feature/keyboard-layout-support/src/Layouts/US_QWERTY.ts
You may like to check #45 for guide of creating layout file.

**Comment by erikbra** [comment]
This is very interesting, @aioutecism. So, I need to just map the same characters as you have in the US layout? I see you are missing an identifier on the key between left shift and z on my keyboard (the bottommost layout has 8 keys between shift and vk_oem_comma. In Norwegian keyboard layout, it is used for <>

**Comment by erikbra** [comment]
Should I make a new pull request with the new layout, btw? What branch should I fork off?

**Comment by aioutecism** [comment]
@erikbra 

The extra key between left shift and z seems to be missing in VSCode's keybinding system.
Cloud you open VSCode's keybindings.json, press `Ctrl+K Ctrl+K` and `The key between left shift and z`, then take a screenshot for me (like the one below). I may find a workaround.

![keybinding_key-binding-widget](https://cloud.githubusercontent.com/assets/2175401/13488155/daee543e-e162-11e5-92ab-2026bedffb72.png)

**Comment by aioutecism** [comment]
@erikbra 
Please checkout a new branch from `feature/keyboard-layout-support`, add the layout file.
Then open a pull request with base: `feature/keyboard-layout-support`, head `your-branch-name`.

**Comment by erikbra** [comment]
It's oem_102.

![image](https://cloud.githubusercontent.com/assets/1628994/13488679/0f55195a-e123-11e5-87cd-14c14160ce08.png)

**Comment by aioutecism** [comment]
Thanks. What would show up in your `keybindings.json` if you press enter after that?

**Comment by erikbra** [comment]
[{
    "key": "oem_102",
    "command": "commandId",
    "when": "editorTextFocus"
}
]

**Comment by aioutecism** [comment]
Thank you.

I added the binding but I'm not sure it will work or not.

https://github.com/aioutecism/amVim-for-VSCode/commit/fae678e80b28cc76c24505a47ea46840825c8a14

Please uncomment-out `'vk_oem_102;: '',` and `shift+vk_oem_102': '',` in your layout file, then fill in the right character.

BTW, you can test on your machine flowing these steps:

1. Clone this repo.
2. cd into root and run `npm install`.
3. Open root folder in VSCode.
4. Create your layout file.
5. Press `F5` to test it out.

**Comment by erikbra** [comment]
Will test, thanks.

**Comment by erikbra** [comment]
Just to be clear, @aioutecism , is it the "right side" of the map that's important, or the actual key (the "left side")? Because e.g. @ is not on any of the keys in the map on my keyboard (it's ctrl+alt+2)

**Comment by aioutecism** [comment]
Please only change the right side of map. 
Leave the left side as they are or the map won't work.

**Comment by erikbra** [comment]
I have a problem with vk_oem_6. On my keyboard, it is a "dead key", i.e. it doesn't map directly to a character, only an accent, which is added to the letter typed after it.

E.g. 
vk_oem6, then e = ë
shift+vk_oem6, then e = ê

etc

So I'm not sure the "virtual keyboard" approach works when having any dead keys.

https://en.wikipedia.org/wiki/Dead_key

**Comment by aioutecism** [comment]
Please comment out `vk_oem_6` and `shift+vk_oem_6`. VSCode should do the rest for us (I hope).

Also, I'm asking a fix to VSCode's keybinding system [here](https://github.com/Microsoft/vscode/issues/3671) if you are interested.

**Comment by aioutecism** [comment]
All layout related problems should be solved in 1.0.0.
However, 1.0.0 depends on VSCode >= 0.10.12.
You can download the VSCode insiders version [here](https://code.visualstudio.com/insiders) and install amVim to try it.

**Comment by erikbra** [comment]
Sorry I haven't commented. But on VSCode 1.1.1, everything works as expected. Nice job!!
